#984 WebSocket Communication in Theia GLSP

ndoschek opened the pull request:
- Use WebSocket communication to GLSP server from Theia backend
  - Extend SocketServerContribution to allow starting server in WebSocket mode
  - Introduce WebSocketConnectionForwarder
  - Add launch configs and scripts to start workflow example in WebSocket mode
- Connect directly to GLSP WebSocket server from frontend
  - Skip binding of backend contribution if `--directWebSocket` argument is passed
  - Open direct WebSocket connection from GLSP frontend client to GLSP WebSocket Server
- Fix logLevel argument
- Update README and CHANGELOG

Resolves eclipse-glsp/glsp/issues/984

tortmayr commented on examples/workflow-theia/src/browser/workflow-glsp-client-contribution.ts:
```diff
@@ -34,4 +38,15 @@ export class WorkflowGLSPClientContribution extends BaseGLSPClientContribution {
+
+    protected override async createGLSPClient(connectionProvider: ConnectionProvider): Promise<GLSPClient> {
+        const envVar = await this.envVariableServer.getValue('WEBSOCKET_PORT');
+        const webSocketPort = envVar && envVar.value ? Number.parseInt(envVar.value, 10) : undefined;
```
We should probably do some error guarding here.
If the websocket envar is defined we should check that its a valid number  (so not NaN or `undefined`) and throw an Error if that's not the case.

tortmayr commented on packages/theia-integration/package.json:
```diff
@@ -44,7 +44,11 @@
   "dependencies": {
-    "@eclipse-glsp/client": "next"
+    "@eclipse-glsp/client": "next",
+    "ws": "^8.13.0"
```
We should probably align this with the versions used in Theia.
I quickly checked and Theia uses:

`"ws:":"~8.11.0"`

tortmayr commented on packages/theia-integration/src/browser/theia-jsonrpc-glsp-client.ts:
```diff
@@ -45,12 +53,30 @@ export class TheiaJsonrpcGLSPClient extends BaseJsonrpcGLSPClient {
+    protected override doCreateConnection(): Promise<MessageConnection> {
+        if (this.webSocketPort) {
+            const websocket = new WebSocket(`ws://localhost:${this.webSocketPort}/${this.id}`);
+            return new Promise(resolve => {
```
You could reuse the `listen` utility function from `@eclipse-glsp/protocol` here

tortmayr commented on packages/theia-integration/src/node/glsp-socket-server-contribution.ts:
```diff
@@ -41,6 +43,9 @@ export interface GLSPSocketServerContributionOptions extends GLSPServerContribut
     additionalArgs?: string[];
+
+    /** Path of the WebSocket GLSP Server. */
+    webSocketPath?: string;
```
I think it would be a nice if we would introduce an abstract `GlspJsonRpcServerLauncher` super class here and then introduce dediacted subclasses for the socket & websocket use case. The orthogonal [PR ](https://github.com/eclipse-glsp/glsp-theia-integration/pull/154) for running the GLSP server directly in the node backend could also use this super class.

ndoschek replied:
That sounds good in general! As discussed with you offline, I will create a follow-up ticket for that.
Because in my use case, I can consume everything from the existing `GLSPSocketServerContribution` and only have a different behaviour in the `connect` function in case of the WebSocket communication.

tortmayr commented on packages/theia-integration/src/node/glsp-socket-server-contribution.ts:
```diff
@@ -50,7 +55,8 @@ export namespace GLSPSocketServerContributionOptions {
                 port: NaN
-            }
+            },
+            webSocketPath: getWebSocketPath()
```
If we introduce dedicated launchers for socket /websocket we could do this flag check in the backend-module and bind the right contribution based on the given parameter set. Similiar to how you do it for  the direct websocket usecase.

tortmayr commented on packages/theia-integration/src/node/websocket-connection-forwarder.ts:
```diff
@@ -0,0 +1,73 @@
+
+    constructor(protected readonly channel: Channel, protected readonly webSocket: WebSocket) {
+        const wrappedWebSocket = wrapWebSocket(webSocket);
+        const reader = new WebSocketMessageReader(wrappedWebSocket);
```
I think we  could simplyfy the forwarding here. It should not be necessary to install reader/writers on top of the websocket. IMO we should be able to simply directly forward the message buffer received from the channel and vice versa.

tortmayr replied:
I think we  could simplify the forwarding here. It should not be necessary to install reader/writers on top of the websocket. IMO we should be able to simply directly forward the message buffer received from the channel and vice versa.

tortmayr reviewed (changes requested):
Thanks for introducing this nice new feature Nina.
I only had a shallow code review for now and have proposed some changes/suggestions as inline comments.
This PR is conflicting with https://github.com/eclipse-glsp/glsp-theia-integration/pull/154, so I will block merging of #154 for now until this is resolved.

ndoschek commented:
Thank you @tortmayr for your first comments. I addressed them and force pushed my commit.
For the remaining issue of the WebSocket to connect duration (which we discussed offline), I introduced a retry for the `InitializeRequest` in the `TheiaJsonrpcGLSPClient`. Let me know what you think about that solution.

tortmayr commented on README.md:
```diff
@@ -73,6 +75,45 @@ cd examples/browser-app
+
+#### Connect to GLSP server from Theia backend via WebSockets
+
+Extend SocketServerContribution to allow starting server in WebSocket mode
```
```suggestion
```

tortmayr commented on README.md:
```diff
@@ -73,6 +75,45 @@ cd examples/browser-app
+#### Connect to GLSP server from Theia backend via WebSockets
+
+Extend SocketServerContribution to allow starting server in WebSocket mode
+Introduce WebSocketConnectionForwarder
```
```suggestion
```

tortmayr reviewed (changes requested):
Thanks for the update. I have a few minor suggestions. I tested all use cases.
The socketm direct websocket and debug usecases now seem to work as expected.
